### ISSUE TITLE
[inductor][perf] Suboptimal codegen for horizontally fused softmax

### ISSUE THREAD

**Issue by jansel** [pytorch/pytorch#129102]
This is an example where the code generated by TorchInductor is suboptimal, and we should be able to do better.

Source code:
```py
@torch.compile
def add_softmax(a, b):
    return torch.softmax(a, -1) + torch.softmax(b, -1)

add_softmax(
    torch.randn((32768, 8192), device="cuda"),
    torch.randn((32768, 8192), device="cuda"),
)
```

Generated code:
```py
@triton.jit
def triton_(in_ptr0, in_ptr1, out_ptr4, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
    xnumel = 32768
    rnumel = 8192
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = tl.full([XBLOCK, RBLOCK], True, tl.int1)
    rbase = tl.arange(0, RBLOCK)[None, :]
    x0 = xindex
    _tmp2 = tl.full([XBLOCK, RBLOCK], float("-inf"), tl.float32)
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tmp0 = tl.load(in_ptr0 + (r1 + (8192*x0)), rmask, eviction_policy='evict_last', other=0.0)
        tmp1 = tl.broadcast_to(tmp0, [XBLOCK, RBLOCK])
        tmp3 = triton_helpers.maximum(_tmp2, tmp1)
        _tmp2 = tl.where(rmask, tmp3, _tmp2)
    tmp2 = triton_helpers.max2(_tmp2, 1)[:, None]
    _tmp8 = tl.full([XBLOCK, RBLOCK], 0, tl.float32)
    _tmp12 = tl.full([XBLOCK, RBLOCK], float("-inf"), tl.float32)
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tmp4 = tl.load(in_ptr0 + (r1 + (8192*x0)), rmask, eviction_policy='evict_last', other=0.0)
        tmp10 = tl.load(in_ptr1 + (r1 + (8192*x0)), rmask, eviction_policy='evict_last', other=0.0)
        tmp5 = tmp4 - tmp2
        tmp6 = tl_math.exp(tmp5)
        tmp7 = tl.broadcast_to(tmp6, [XBLOCK, RBLOCK])
        tmp9 = _tmp8 + tmp7
        _tmp8 = tl.where(rmask, tmp9, _tmp8)
        tmp11 = tl.broadcast_to(tmp10, [XBLOCK, RBLOCK])
        tmp13 = triton_helpers.maximum(_tmp12, tmp11)
        _tmp12 = tl.where(rmask, tmp13, _tmp12)
    tmp8 = tl.sum(_tmp8, 1)[:, None]
    tmp12 = triton_helpers.max2(_tmp12, 1)[:, None]
    _tmp18 = tl.full([XBLOCK, RBLOCK], 0, tl.float32)
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tmp14 = tl.load(in_ptr1 + (r1 + (8192*x0)), rmask, eviction_policy='evict_last', other=0.0)
        tmp15 = tmp14 - tmp12
        tmp16 = tl_math.exp(tmp15)
        tmp17 = tl.broadcast_to(tmp16, [XBLOCK, RBLOCK])
        tmp19 = _tmp18 + tmp17
        _tmp18 = tl.where(rmask, tmp19, _tmp18)
    tmp18 = tl.sum(_tmp18, 1)[:, None]
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tmp20 = tl.load(in_ptr0 + (r1 + (8192*x0)), rmask, eviction_policy='evict_first', other=0.0)
        tmp24 = tl.load(in_ptr1 + (r1 + (8192*x0)), rmask, eviction_policy='evict_first', other=0.0)
        tmp21 = tmp20 - tmp2
        tmp22 = tl_math.exp(tmp21)
        tmp23 = tmp22 / tmp8
        tmp25 = tmp24 - tmp12
        tmp26 = tl_math.exp(tmp25)
        tmp27 = tmp26 / tmp18
        tmp28 = tmp23 + tmp27
        tl.store(out_ptr4 + (r1 + (8192*x0)), tmp28, rmask)
```

Logically, this code look suboptimal.  We *should* be able to do this kernel in 3 loops, but instead inductor generated 4 loops:
1) max_a
2) max_b, sum_a
3) sum_b
4) output

Where I was expecting it to generate:
1) max_a, max_b 
2) sum_a, sum_a
3) output

The perf results are a bit strange on my local RTX 3090:
- The original 4-loop version took 0.008320 s.
- If I move the max_b to the first loop: 0.007362 s (a 12% speedup)
- The 3-loop version takes: 0.008592 s (3% slower)
- I also tried a 5 loop version and got: 0.007437 s

Patch to go to the fastest version (moving the second max into the first loop): 
```diff
--- a/softmax_output_triton.py
+++ b/softmax_output_triton.py
@@ -42,6 +42,7 @@ def triton_(in_ptr0, in_ptr1, out_ptr4, xnumel, rnumel, XBLOCK : tl.constexpr, R
     rbase = tl.arange(0, RBLOCK)[None, :]
     x0 = xindex
     _tmp2 = tl.full([XBLOCK, RBLOCK], float("-inf"), tl.float32)
+    _tmp12 = tl.full([XBLOCK, RBLOCK], float("-inf"), tl.float32)
     for roffset in range(0, rnumel, RBLOCK):
         rindex = roffset + rbase
         rmask = rindex < rnumel
@@ -50,25 +51,24 @@ def triton_(in_ptr0, in_ptr1, out_ptr4, xnumel, rnumel, XBLOCK : tl.constexpr, R
         tmp1 = tl.broadcast_to(tmp0, [XBLOCK, RBLOCK])
         tmp3 = triton_helpers.maximum(_tmp2, tmp1)
         _tmp2 = tl.where(rmask, tmp3, _tmp2)
+        tmp10 = tl.load(in_ptr1 + (r1 + (8192*x0)), rmask, eviction_policy='evict_last', other=0.0)
+        tmp11 = tl.broadcast_to(tmp10, [XBLOCK, RBLOCK])
+        tmp13 = triton_helpers.maximum(_tmp12, tmp11)
+        _tmp12 = tl.where(rmask, tmp13, _tmp12)
     tmp2 = triton_helpers.max2(_tmp2, 1)[:, None]
+    tmp12 = triton_helpers.max2(_tmp12, 1)[:, None]
     _tmp8 = tl.full([XBLOCK, RBLOCK], 0, tl.float32)
-    _tmp12 = tl.full([XBLOCK, RBLOCK], float("-inf"), tl.float32)
     for roffset in range(0, rnumel, RBLOCK):
         rindex = roffset + rbase
         rmask = rindex < rnumel
         r1 = rindex
         tmp4 = tl.load(in_ptr0 + (r1 + (8192*x0)), rmask, eviction_policy='evict_last', other=0.0)
-        tmp10 = tl.load(in_ptr1 + (r1 + (8192*x0)), rmask, eviction_policy='evict_last', other=0.0)
         tmp5 = tmp4 - tmp2
         tmp6 = tl_math.exp(tmp5)
         tmp7 = tl.broadcast_to(tmp6, [XBLOCK, RBLOCK])
         tmp9 = _tmp8 + tmp7
         _tmp8 = tl.where(rmask, tmp9, _tmp8)
-        tmp11 = tl.broadcast_to(tmp10, [XBLOCK, RBLOCK])
-        tmp13 = triton_helpers.maximum(_tmp12, tmp11)
-        _tmp12 = tl.where(rmask, tmp13, _tmp12)
     tmp8 = tl.sum(_tmp8, 1)[:, None]
-    tmp12 = triton_helpers.max2(_tmp12, 1)[:, None]
     _tmp18 = tl.full([XBLOCK, RBLOCK], 0, tl.float32)
     for roffset in range(0, rnumel, RBLOCK):
         rindex = roffset + rbase
```

Patch to fuse the two sum loops together (makes thing slower):
```diff
diff --git a/benchmarks/dynamo/plot.py b/benchmarks/dynamo/plot.py
deleted file mode 100644
index e69de29bb2d..00000000000
diff --git a/softmax_output_triton.py b/softmax_output_triton.py
index 22f028266de..bbc53e4b096 100644
--- a/softmax_output_triton.py
+++ b/softmax_output_triton.py
@@ -58,6 +58,7 @@ def triton_(in_ptr0, in_ptr1, out_ptr4, xnumel, rnumel, XBLOCK : tl.constexpr, R
     tmp2 = triton_helpers.max2(_tmp2, 1)[:, None]
     tmp12 = triton_helpers.max2(_tmp12, 1)[:, None]
     _tmp8 = tl.full([XBLOCK, RBLOCK], 0, tl.float32)
+    _tmp18 = tl.full([XBLOCK, RBLOCK], 0, tl.float32)
     for roffset in range(0, rnumel, RBLOCK):
         rindex = roffset + rbase
         rmask = rindex < rnumel
@@ -68,18 +69,13 @@ def triton_(in_ptr0, in_ptr1, out_ptr4, xnumel, rnumel, XBLOCK : tl.constexpr, R
         tmp7 = tl.broadcast_to(tmp6, [XBLOCK, RBLOCK])
         tmp9 = _tmp8 + tmp7
         _tmp8 = tl.where(rmask, tmp9, _tmp8)
-    tmp8 = tl.sum(_tmp8, 1)[:, None]
-    _tmp18 = tl.full([XBLOCK, RBLOCK], 0, tl.float32)
-    for roffset in range(0, rnumel, RBLOCK):
-        rindex = roffset + rbase
-        rmask = rindex < rnumel
-        r1 = rindex
         tmp14 = tl.load(in_ptr1 + (r1 + (8192*x0)), rmask, eviction_policy='evict_last', other=0.0)
         tmp15 = tmp14 - tmp12
         tmp16 = tl_math.exp(tmp15)
         tmp17 = tl.broadcast_to(tmp16, [XBLOCK, RBLOCK])
         tmp19 = _tmp18 + tmp17
         _tmp18 = tl.where(rmask, tmp19, _tmp18)
+    tmp8 = tl.sum(_tmp8, 1)[:, None]
     tmp18 = tl.sum(_tmp18, 1)[:, None]
     for roffset in range(0, rnumel, RBLOCK):
         rindex = roffset + rbase
```

@eellison @shunting314 this codegen might be something we should try to fix on the inductor side.  Since splitting loops up resulted in speedups, maybe we should explore doing some kernels in more loops.

@bertmaher @htyu this more loops = faster perf behavior of Triton seems a bit odd to me. Perhaps there is something on the Triton side we could improve?  Maybe digging into that 12% could yield some improvement opportunities.

cc @ezyang @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire

**Comment by shunting314** [comment]
> this codegen might be something we should try to fix on the inductor side. Since splitting loops up resulted in speedups, maybe we should explore doing some kernels in more loops.

yea. I think maybe doing more things in a loop results in more register pressures and cause perf problem. Will need verify with some profiling though

**Comment by jansel** [comment]
Maybe we should actually do 5 loops here:
1) max_a
2) sum_a (in reverse order to maximize cache hits)
3) max_b
4) sum_b (in reverse order to maximize cache hits)
5) output

**Comment by htyu** [comment]
> yea. I think maybe doing more things in a loop results in more register pressures and cause perf problem. Will need verify with some profiling though

It's very likely. With two separated loops, the intermediate result `_tmp8` will be stored to SMEM via `tl.sum` to free some regsiters, which isn't too bad (much better than tl.store).

**Comment by htyu** [comment]
> Maybe we should actually do 5 loops here:
> 
> 1. max_a
> 2. sum_a (in reverse order to maximize cache hits)
> 3. max_b
> 4. sum_b (in reverse order to maximize cache hits)
> 5. output

This sounds reasonable to me. In general I would not fuse loops that do not have data dependencies into one. When they have data dependencies, fusion can be helpful in that the intermediate results from the first loop can be pipelined directly to the second loop in the same loop iteration. When there are no dependencies, loop fission (loop distribution) would be the way to go usually.
